sql/core/src/main/scala/org/apache/spark/sql/columnar/ColumnAccessor.scala

```diff
@@ -117,9 +135,14 @@ private[sql] object ColumnAccessor {
       case DoubleType => new DoubleColumnAccessor(buf)
       case StringType => new StringColumnAccessor(buf)
       case BinaryType => new BinaryColumnAccessor(buf)
-      case DecimalType.Fixed(precision, scale) if precision < 19 =>
-        new FixedDecimalColumnAccessor(buf, precision, scale)
-      case other => new GenericColumnAccessor(buf, other)
+      case dt: DecimalType if dt.precision < 19 => new CompactDecimalColumnAccessor(buf, dt)
```

Comment:
in other places we use `<= Decimal.MAX_LONG_DIGITS`, I think we should follow it here.